REGRESSION(271419@main): [GTK][WPE] Build fails if clang <= 14

dpino opened the pull request:
#### 0a646f68779a3c4e259394fcf3e64ba8d22b7272
<pre>
REGRESSION(271419@main): [GTK][WPE] Build fails if clang &lt;= 14
<a href="https://bugs.webkit.org/show_bug.cgi?id=266244">https://bugs.webkit.org/show_bug.cgi?id=266244</a>

Reviewed by Philippe Normand.

This is a partial revert of 271419@main to avoid using
std::views:values, which is not supported by clang-14.

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::flush):
(WebCore::SourceBufferPrivateGStreamer::enqueueSample):
(WebCore::SourceBufferPrivateGStreamer::isReadyForMoreSamples):
(WebCore::SourceBufferPrivateGStreamer::notifyClientWhenReadyForMoreSamples):
(WebCore::SourceBufferPrivateGStreamer::allSamplesInTrackEnqueued):
(WebCore::SourceBufferPrivateGStreamer::precheckInitializationSegment):
(WebCore::SourceBufferPrivateGStreamer::platformMaximumBufferSize const):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:

Canonical link: <a href="https://commits.webkit.org/271921@main">https://commits.webkit.org/271921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5ce1dfa4b56c69fe2138c82e6f56b39665861a6f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/8641 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/31293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/32485 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27104 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/30636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5896 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27207 "Found 58 new test failures: http/tests/media/media-source/media-source-video-fit-fill.html, http/tests/media/media-source/mediasource-play-then-seek-back-with-remote-control.html, http/tests/media/media-source/mediasource-rvfc.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-events.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-multisession.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-setMediaKeys-after-update.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-setMediaKeys-onencrypted.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-waiting-for-a-key.https.html, imported/w3c/web-platform-tests/media-source/mediasource-activesourcebuffers.html, imported/w3c/web-platform-tests/media-source/mediasource-append-buffer.html ... (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/30270 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/7243 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25569 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6185 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/6342 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26658 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33821 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/27338 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27090 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32589 "Found 67 new test failures: fast/canvas/webgl/texImage2D-mse-flipY-false.html, fast/canvas/webgl/texImage2D-mse-flipY-true.html, fast/canvas/webgl/texImage2D-mse-in-dom-flipY-false.html, fast/canvas/webgl/texImage2D-mse-in-dom-flipY-true.html, http/tests/media/media-source/media-source-video-fit-fill.html, http/tests/media/media-source/mediasource-play-then-seek-back-with-remote-control.html, http/tests/media/media-source/mediasource-rvfc.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-clear-encrypted.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-encrypted-clear.https.html, imported/w3c/web-platform-tests/encrypted-media/clearkey-mp4-playback-temporary-events.https.html ... (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/6288 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/4459 "Found 1 new test failure: fullscreen/full-screen-layer-dump.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/30316 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8026 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7123 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7031 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
```diff
@@ -209,21 +209,21 @@ bool SourceBufferPrivateGStreamer::precheckInitializationSegment(const Initializ
         GRefPtr<GstCaps> initialCaps = videoTrackInfo->initialCaps();
         ASSERT(initialCaps);
         if (!m_tracks.contains(trackInfo.track->id()))
-            m_tracks.try_emplace(trackInfo.track->id(), MediaSourceTrackGStreamer::create(TrackPrivateBaseGStreamer::TrackType::Video, trackInfo.track->id(), videoTrackInfo->stringId(), WTFMove(initialCaps)));
+            m_tracks.add(trackInfo.track->id(), MediaSourceTrackGStreamer::create(TrackPrivateBaseGStreamer::TrackType::Video, trackInfo.track->id(), videoTrackInfo->stringId(), WTFMove(initialCaps)));
```
Here `HashMap::ensure()` can be used maybe?

dpino replied:
Originally the code was using `m_tracks.add`.

https://github.com/WebKit/WebKit/pull/20901/files#diff-cbb52de8c051ac0defaf4bb8ed562696b3a14327efdcb067be7b23e912c432d9L219

Nevertheless, I tried using `m_tracks.ensure` but I got a build error:

```bash
WTF/Headers/wtf/HashMap.h:246:65: error: type 'WTF::Ref<WebCore::MediaSourceTrackGStreamer>' does not provide a call operator
        ValueTraits::ValueTraits::assignToEmpty(location.value, functor());
                                                                ^~~~~~~
WTF/Headers/wtf/HashTable.h:938:25: note: in instantiation of function template specialization 'WTF::HashMapEnsureTranslator<WTF::HashMap<unsigned long, WTF::RefPtr<WebCore::MediaSourceTrackGStreamer>>::KeyValuePairTraits, WTF::DefaultHash<unsigned long>>::translate<WTF::Key
        HashTranslator::translate(*entry, std::forward<T>(key), std::forward<Extra>(extra));
                        ^
WTF/Headers/wtf/HashMap.h:402:28: note: in instantiation of function template specialization 'WTF::HashTable<unsigned long, WTF::KeyValuePair<unsigned long, WTF::RefPtr<WebCore::MediaSourceTrackGStreamer>>, WTF::KeyValuePairKeyExtractor<WTF::KeyValuePair<unsigned long, WTF::
    return m_impl.template add<HashMapEnsureTranslator<KeyValuePairTraits, HashFunctions>>(std::forward<K>(key), std::forward<F>(functor));
                           ^
WTF/Headers/wtf/HashMap.h:465:12: note: in instantiation of function template specialization 'WTF::HashMap<unsigned long, WTF::RefPtr<WebCore::MediaSourceTrackGStreamer>>::inlineEnsure<unsigned long, WTF::Ref<WebCore::MediaSourceTrackGStreamer>>' requested here
    return inlineEnsure(std::forward<KeyType>(key), std::forward<Functor>(functor));
           ^
../../../Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:225:22: note: in instantiation of function template specialization 'WTF::HashMap<unsigned long, WTF::RefPtr<WebCore::MediaSourceTrackGStreamer>>::ensure<WTF::Ref<WebCore::MediaSourceTrac
            m_tracks.ensure(trackInfo.track->id(), MediaSourceTrackGStreamer::create(TrackPrivateBaseGStreamer::TrackType::Video, trackInfo.track->id(), videoTrackInfo->stringId(), WTFMove(initialCaps)));
                     ^
1 error generated.
```

philn replied:
Well, the second argument has to be a lambda...

Anyways, whatever. r+

philn commented on Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
```diff
@@ -209,21 +209,21 @@ bool SourceBufferPrivateGStreamer::precheckInitializationSegment(const Initializ
         GRefPtr<GstCaps> initialCaps = audioTrackInfo->initialCaps();
         ASSERT(initialCaps);
         if (!m_tracks.contains(trackInfo.track->id()))
-            m_tracks.try_emplace(trackInfo.track->id(), MediaSourceTrackGStreamer::create(TrackPrivateBaseGStreamer::TrackType::Audio, trackInfo.track->id(), audioTrackInfo->stringId(), WTFMove(initialCaps)));
```
Ditto, and below as well

dpino committed:
REGRESSION(271419@main): [GTK][WPE] Build fails if clang <= 14
https://bugs.webkit.org/show_bug.cgi?id=266244

Reviewed by Philippe Normand.

This is a partial revert of 271419@main to avoid using
std::views:values, which is not supported by clang-14.

* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.cpp:
(WebCore::SourceBufferPrivateGStreamer::flush):
(WebCore::SourceBufferPrivateGStreamer::enqueueSample):
(WebCore::SourceBufferPrivateGStreamer::isReadyForMoreSamples):
(WebCore::SourceBufferPrivateGStreamer::notifyClientWhenReadyForMoreSamples):
(WebCore::SourceBufferPrivateGStreamer::allSamplesInTrackEnqueued):
(WebCore::SourceBufferPrivateGStreamer::precheckInitializationSegment):
(WebCore::SourceBufferPrivateGStreamer::platformMaximumBufferSize const):
* Source/WebCore/platform/graphics/gstreamer/mse/SourceBufferPrivateGStreamer.h:

Canonical link: https://commits.webkit.org/271921@main
